### DISCUSSION TITLE
Using empty scoped policies for multitenancy

### DISCUSSION THREAD

**Discussion by haines** [cerbos/cerbos#688] [Ideas]
Suppose I run a platform with a large number of tenants. Most tenants have no custom authz requirements, but some do, so I need to use scoped policies.

My application shouldn't need to have knowledge of whether a tenant has custom authz, so it will unconditionally pass the tenant identifier as a scope along with every resource.

Currently, I have to define a policy for every tenant+resource pair, and I can't leave those policies empty so I have to add dummy rules that will never match.

At a minimum, I'd like to be able to leave the policies empty for the tenants that should just inherit the base set of rules. I think this might also make sense in cases where the parent doesn't want to define any rules but rather leave it up to the children.

However, it might be nicer to avoid having to define the empty policies at all. Note that I would want to keep the default behaviour for the tenants that _do_ have custom authz requirements, so that if I introduce a new resource policy without the scoped versions, all actions will be denied on those resources for those scopes. This implies being able to set the missing-policy behaviour per scope - either "treat as empty" or "deny all" (which would still be the default).

So, to summarise, two questions:

1. Could we relax the validation requirement to allow `rules` to be empty? 
   **Yes** - implemented in #694 and released in v0.14.0
2. Could we allow missing policies to be treated as empty for specific scopes?
   **No** - but we might want to look at tooling to make it easier to create the empty policies.

**Comment by charithe** [comment]
I was thinking about this myself. I think that if the scope requested by the API call does not exist, we should try descending down the scope hierarchy until we find a match.

**Reply by haines** [reply]
I think in some cases you would want the stricter behaviour.

Example: a tenant is internally structured into regions (APAC and EMEA), so the policies for that tenant deny actions when the principal's region does not match the resource's region.

If I introduce a new resource to the system, I wouldn't want to fall back to the parent scope, because that would mean I accidentally allow the APAC users to access EMEA resources until I add the scoped policy.

For that reason I think the behaviour should be configurable and we should default to the stricter setting.

**Reply by haines** [reply]
As for _how_ to configure it, perhaps we would introduce a new policy type, something like

```yaml
apiVersion: api.cerbos.dev/v1
scopePolicy:
  scope: com.foo.bar.inherit
  missingPolicies: EFFECT_NONE # or EFFECT_DENY
```

**Reply by charithe** [reply]
I think it would be better if it was just a simple configuration flag set in the PDP config file. This doesn't feel like something that deserves a policy kind of its own. If we did do it, then there'll be a bunch of complications to handle on our side like what to do if there is more than one policy defined for the same scope, what to do with conflicting policies defined for ancestors, invalidating caches for all affected policies when one of these is updated and so on. It'd be complicated for the users as well because these things will be interspersed among other policies. Finding them and reasoning about them will be harder without special tooling.

**Reply by haines** [reply]
Hmm, I really feel like it's data rather than config. It seems awkward to have to reconfigure and restart the PDP to change this behaviour when I add a new tenant to the system (normally this would only require me to introduce a bunch of principal/resource policies with `scope` set to the new tenant's value). Also, if we offered a hosted PDP, how would we allow users to set it?

>what to do if there is more than one policy defined for the same scope

I'd expect that to handled as a `DuplicateDef` error, the same as for the other policy types.

>what to do with conflicting policies defined for ancestors

My original thought was that this only applies to the scope itself, so that setting `missingPolicies` for `com.foo` doesn't affect `com.foo.bar`. We could also specify the value is inherited, so that if `missingPolicies` is set for `com.foo` it also applies to `com.foo.bar`, unless `com.foo.bar` also sets the value.

>invalidating caches for all affected policies when one of these is updated

Yeah, this is slightly painful. Presumably we would have to have an index into the cache that maps scope to cache entries?

>It'd be complicated for the users as well because these things will be interspersed among other policies

Perhaps we could guide users towards a directory structure like

```
policies/
  principal/
  resource/
  scope/
```

**Reply by charithe** [reply]
What's the advantage of setting this behaviour on a scope-by-scope basis? In my mind, it's a simple binary choice: either you want to be strict and fail if the requested scope doesn't exist or you want to be lenient and try descending into the ancestors.  I don't see why the value of the scope matters.

**Reply by haines** [reply]
Suppose I have two tenants (/scopes) called `vanilla` and `regional`. 

`vanilla` is a simple tenant who has no custom authz requirements. I don't want to have to define empty policies for every resource for this tenant, so I create a scope policy and set `missingPolicies: EFFECT_NONE` so that I fall back to the root-scoped policies.

`regional` is a complex tenant with custom authz requirements (users are divided into APAC and EMEA, and can't access each others' resources). If I introduce a new resource policy at the root scope, I must ensure that I've defined a `regional`-scoped policy for that resource as well, otherwise I'll be allowing the users to access resources from the other region. Therefore, I want `missingPolicies: EFFECT_DENY` for this scope (which is the default, or I can explicitly set it with a scope policy), so that I don't accidentally leak data.

**Reply by dbuduev** [reply]
If I have understood correctly, the `regional` tenant doesn't want the EMEA and APAC scopes to inherit any rule from the root scope. 
If inheritance isn't needed, the tenant can create isolated scope hierarchies, so EMEA and APAC scope won't have a common ancestor.

**Reply by charithe** [reply]
I am looking at this from the point of view of the operator of the SaaS. I have a set of  no-scope policies that define the minimal set of access rules for the resources in the system. I onboard a new tenant called `acme`. If I am on "strict" mode, I must create (possibly empty) copies of the resource policies and set their scope to `acme` or the tenant won't be able to do anything at all. If I am on "lenient" mode, I don't have to make those copies eagerly. I can wait until the tenant requires special permissions for particular resources and create scoped policies for just those.

Now, the tenant might want further customization based on teams, departments etc. Let's say they create a `marketing` team (scope=`acme.marketing`). Again, in strict mode, there must be (possibly empty) copies of policies with scope `acme.marketing` in order for the members of the marketing team to do anything. In lenient mode they don't.

Obviously all of this copying of policies should be done automatically via the provisioning system so it's not visible to users (they might not even be aware of scopes at all). So, assuming that we don't directly expose policy management to end users, isn't it easier to manage when strict/lenient is a global setting?

**Reply by haines** [reply]
There's a worked example of the scenario I'm envisaging here: https://github.com/cerbos/demo-multitenant-saas

Thinking about operating the SaaS platform, I've got to consider onboarding a new tenant but also making changes that affect all tenants like adding a new resource kind.

I wouldn't want global settings when I onboard a new tenant. If the new tenant has no custom authz requirements, then I just want to set that tenant to "lenient" mode to use the set of no-scope policies. However, if I onboard a tenant that does have custom authz, then leaving that tenant in "strict" mode is valuable because it forces me to consider those custom authz requirements across the whole system - I can't forget to add a scoped policy for a resource and thereby accidentally grant wider permissions than I intended.

Apart from onboarding, I also have to think about what happens when I release a feature that adds a new kind of resource to my system. I also wouldn't want a global setting here, because with per-scope settings I get the benefit that
- My lenient tenants will automatically inherit the base set of permissions when I add the no-scope policy for the new resource kind.
- To roll out my new feature to the strict tenants, I am forced to implement the scoped policies before the feature can be accessed - again, avoiding me accidentally granting too many permissions.

Trying to think about it from another angle, perhaps it's a problem that could be potentially solved with tooling rather introducing a strict/lenient setting at all - maybe the underlying issue is just that "it's a bit of a pain that I have to define empty policies to get a scope to inherit from its parent", if we made it easier to create those policies then we could just leave the existing strict behaviour.

If I could run a command that would take a scope, find all no-scope policies, and generate corresponding (empty) scoped policies (if one does not exist already), then maybe that's sufficent?

**Reply by dbuduev** [reply]
I agree that it's beneficial to have fine-grained control over the mode of inheritance ("strict" or "lenient").

> Trying to think about it from another angle, perhaps it's a problem that could be potentially solved with tooling rather introducing a strict/lenient setting at all - maybe the underlying issue is just that "it's a bit of a pain that I have to define empty policies to get a scope to inherit from its parent", if we made it easier to create those policies then we could just leave the existing strict behaviour.

I like this approach.

Alternatively, we can implement a lenient mode behaviour for scoped policies and model a strict mode by other means (see below).

In this case as a SaaS operator I can:
1. Assign a `root` scope to a base set of policy, so there are no no-scope policies.
2. If I want a lenient mode for `acme` tenant, I assign value `root.acme` to its scope.
3. If I want a strict mode for `acme`, I assign it `root-acme` to its scope. Here I create a new scope hierarchy.

My thinking is based on assumption that strict mode means - no inheritance.

**Reply by charithe** [reply]
I think we are kind of on the same page :-)  My point was that it would be a bit more complicated to reason about the system if we had this added dimension of strict/lenient setting for each scope. I agree it's a pain to create empty policies to make a feature available to someone when you are in strict mode. Tooling is definitely required. My assumption was that it would be built by the operator when they designed the system (or by us). Even though having empty policies is not very elegant, it's very explicit. When you look at someone's set of policies you immediately know what features they have access to.

 
> My thinking is based on assumption that strict mode means - no inheritance.

@dbuduev this is about request handling. In the API request (`check`) you have to define the scope your request belongs to. Usually it would be your position in the hierarchy like `acme.marketing.uk`. In strict mode, there must be a policy (potentially empty) with the same scope or your request will be denied. In lenient mode, if we don't find a policy with `acme.marketing.uk`, instead of failing, we try `acme.marketing`, `acme` and `` as well. There's still fall-through to ancestors during policy evaluation. This is about locating the policy to evaluate in the first place.
